url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/triggers/SimpleTriggerImpl.java

```diff
@@ -0,0 +1,605 @@
+   * </p>
+   * <p>
+   * After this method has been called, <code>getNextFireTime()</code> should return a valid answer.
+   * </p>
+   *
+   * @return the first time at which the <code>Trigger</code> will be fired by the scheduler, which is also the same value
+   *         <code>getNextFireTime()</code> will return (until after the first firing of the <code>Trigger</code>).
+   *         </p>
+   */
+  @Override
+  public Date computeFirstFireTime(Calendar calendar) {
+
+    nextFireTime = getStartTime();
+
+    while (nextFireTime != null && calendar != null && !calendar.isTimeIncluded(nextFireTime.getTime())) {
+      nextFireTime = getFireTimeAfter(nextFireTime);
+
+      if (nextFireTime == null) {
+        break;
+      }
+
+      // avoid infinite loop
+      java.util.Calendar c = java.util.Calendar.getInstance();
+      c.setTime(nextFireTime);
+      if (c.get(java.util.Calendar.YEAR) > YEAR_TO_GIVEUP_SCHEDULING_AT) {
+        return null;
+      }
+    }
+
+    return nextFireTime;
+  }
+
+  /**
+   * <p>
+   * Returns the next time at which the <code>Trigger</code> is scheduled to fire. If the trigger will not fire again, <code>null</code> will be
+   * returned. Note that the time returned can possibly be in the past, if the time that was computed for the trigger to next fire has already
+   * arrived, but the scheduler has not yet been able to fire the trigger (which would likely be due to lack of resources e.g. threads).
+   * </p>
+   * <p>
+   * The value returned is not guaranteed to be valid until after the <code>Trigger</code> has been added to the scheduler.
+   * </p>
+   *
+   * @see TriggerUtils#computeFireTimesBetween(Trigger, Calendar, Date, Date)
+   */
+  @Override
+  public Date getNextFireTime() {
+
+    return nextFireTime;
+  }
+
+  /**
+   * <p>
+   * Returns the previous time at which the <code>SimpleTrigger</code> fired. If the trigger has not yet fired, <code>null</code> will be returned.
+   */
+  @Override
+  public Date getPreviousFireTime() {
+
+    return previousFireTime;
+  }
+
+  /**
+   * <p>
+   * Set the next time at which the <code>SimpleTrigger</code> should fire.
+   * </p>
+   * <p>
+   * <b>This method should not be invoked by client code.</b>
+   * </p>
+   */
+  @Override
+  public void setNextFireTime(Date nextFireTime) {
+
+    this.nextFireTime = nextFireTime;
+  }
+
+  /**
+   * <p>
+   * Set the previous time at which the <code>SimpleTrigger</code> fired.
+   * </p>
+   * <p>
+   * <b>This method should not be invoked by client code.</b>
+   * </p>
+   */
+  @Override
+  public void setPreviousFireTime(Date previousFireTime) {
+
+    this.previousFireTime = previousFireTime;
+  }
+
+  /**
+   * <p>
+   * Returns the next time at which the <code>SimpleTrigger</code> will fire, after the given time. If the trigger will not fire after the given time,
+   * <code>null</code> will be returned.
+   * </p>
+   */
+  @Override
+  public Date getFireTimeAfter(Date afterTime) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 478-514 into a separate method.